### PR TITLE
Refactor JsonPatch internals

### PR DESCRIPTION
**Pull request by gcanti**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved JSON Patch generation and application for arrays, objects, and root-level operations.
  * Added stricter validation for invalid JSON Pointer paths, unsupported root `remove`, and array index handling (including correct `-` usage).
  * Properly treats special keys like `__proto__` as normal own properties during patch generation and application.
  * Preserves reference identity for unaffected subtrees and does not mutate the original document.
* **Tests**
  * Expanded coverage for edge cases, error messages, immutability, inherited-properties rejection, and root-level `add` behavior.

### CODE REVIEW

File: packages/effect/src/JsonPatch.ts [hunk 1]
```diff
@@ -339,72 +317,44 @@ function toIndex(token: string): number {
...
     const updated = { ...parent }
-    if (mode === "remove") delete updated[lastToken]
-    else InternalRecord.assignProperty(updated, lastToken, val!)
+    if (op.op === "remove") delete updated[lastToken]
+    else InternalRecord.assignProperty(updated, lastToken, op.value)
     return rebuildFromStack(stack, updated)
   }
 
-  throw new Error(`Cannot ${mode} at "${pointer}" (parent not found or not a container).`)
+  throw new Error(`Cannot ${op.op} at "${op.path}" (parent not found or not a container).`)
```
[the remark this answers is not in the document]
Comment (gcanti) [reply packages/effect/src/JsonPatch.ts:357]: Invalid array index is pre-existing observable behavior, already covered for final tokens. Normalizing it to Cannot ... at would be a behavioral regression. I added coverage for intermediate array tokens to make the preserved contract explicit.
